### PR TITLE
optimizations in compilation and threading

### PR DESCRIPTION
**Pull request by SteveMacenski**
Partially to clean things up but also compiling with some of these flags seems to help and are the same compiler flags as used in https://romanpoya.medium.com/a-look-at-the-performance-of-expression-templates-in-c-eigen-vs-blaze-vs-fastor-vs-armadillo-vs-2474ed38d982 for comparison eith eigen/xtensor/fastor/etc more or less just assuming they know what they're doing. 

While the system gets into the 20s of ms, it spends more time in the teens, so I think it has incremental help. I didn't do formal benchmarking, but its clear from the stream of values coming back to me that these flags improved things. 

I haven't looked into it yet, but most of the time that we miss our loop rate, its due to 1 really odd cycle that is high. I suspect that's due to the path handler, so that might be an area to consider for optimization.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,13 @@ foreach(pkg IN LISTS dependencies_pkgs)
 endforeach()
 
 nav2_package()
+add_compile_options(-O3 -mavx2 -mfma -finline-limit=1000000 -ffp-contract=fast)
+
+include_directories(
+  include
+  ${xsimd_INCLUDE_DIRS}
+  ${OpenMP_INCLUDE_DIRS}
+)
```
Comment (artofnothingness) [inline CMakeLists.txt]: target approach more preferred. Why do we need this ?

Comment (SteveMacenski) [reply CMakeLists.txt]: Every file uses xtensor and wants xsimd or openmp optimizations. These seem like a rather global thing to me. Its not strictly required.

**Comment by padhupradheep** [comment]
Off topic and a very basic question, (sorry)

why aren't we setting a ``CMAKE_BUILD_TYPE`` , because as far as I understand, there is no default set. For a much better optimized performance we have to use the ``RELEASE``  flag right ? Or is it set somewhere else ?

https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html

File: CMakeLists.txt (hunk 1 above)
Comment (artofnothingness) [reply CMakeLists.txt]: let's stick to target approach

**Comment by artofnothingness** [comment]
> Off topic and a very basic question, (sorry)
> 
> why aren't we setting a `CMAKE_BUILD_TYPE` , because as far as I understand, there is no default set. For a much better optimized performance we have to use the `RELEASE` flag right ? Or is it set somewhere else ?
> 
> https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html

it usually comes as an argument

**Comment by SteveMacenski** [comment]
`nav2_package()` macro handles that stuff, and sets to release if nothing is specified

```
  if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
    message(STATUS "Setting build type to Release as none was specified.")
    set(CMAKE_BUILD_TYPE "Release" CACHE
        STRING "Choose the type of build." FORCE)
    # Set the possible values of build type for cmake-gui
    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS
      "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
  endif()
```

Its also where C++17 is set and other related compiler warning flags


```
  # Default to C++14
  if(NOT CMAKE_CXX_STANDARD)
    set(CMAKE_CXX_STANDARD 17)
  endif()

  if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
    add_compile_options(-Wall -Wextra -Wpedantic -Werror -Wdeprecated -fPIC)
  endif()
```

Among other things

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,8 +1,8 @@
 cmake_minimum_required(VERSION 3.5)
 project(mppic)
 
-set(XTENSOR_USE_XSIMD 1)
-set(XTENSOR_USE_OPENMP 1)
+add_definitions(-DXTENSOR_ENABLE_XSIMD)
+add_definitions(-DXTENSOR_USE_XSIMD)
```
Comment (artofnothingness) [inline CMakeLists.txt:5]: What was the purpose of this change ?

File: CMakeLists.txt [hunk 3]
```diff
@@ -53,7 +55,7 @@ set(libraries mppic critics)
 
 foreach(lib IN LISTS libraries)
   target_compile_options(${lib} PUBLIC -fconcepts)
-  target_include_directories(${lib} PUBLIC include)
+  target_include_directories(${lib} PUBLIC include ${xsimd_INCLUDE_DIRS} ${OpenMP_INCLUDE_DIRS})
```
Comment (artofnothingness) [inline CMakeLists.txt:58]: Is it required ?

Comment (SteveMacenski) [reply CMakeLists.txt:58]: Yes, it seems so

File: CMakeLists.txt (hunk 2 above)
Comment (SteveMacenski) [reply CMakeLists.txt:5]: Honestly, mostly paranoia. I couldn't see any changes with or without `XTENSOR_USE_OPENMP` / `XTENSOR_USE_XSIMD`. I don't know if (and certainly it shouldn't be the case that) calling it as a definition would help, but this is how I see it done in the guides for xtensor when looking into optimization options for compilation. So I didn't see a reason to argue with what others are doing.

**Pull request merged**